### PR TITLE
feat: detect completed txs and throw error on fail

### PR DESCRIPTION
**Pull request by elmutt**
Detect completed & failed transactions.

Will have a followup pr to translate error toast on trade error on web

To test:

link to web. enable thor. trade into a coin that completed quickly (eth to ltc is good) and see the trade spinner until it completes

### CODE REVIEW

File: packages/swapper/src/swappers/thorchain/types.ts [hunk 1]
```diff
@@ -21,6 +21,21 @@ export type PoolResponse = {
...
+type action = {
+  date: string
+  height: string
+  out: [out]
+  status: string
+  type: string
+}
+
+export type ActionsResponse = {
+  actions: [action]
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/thorchain/types.ts:36]: `T[]` (preferred for consistency) or `Array<T>` 🙏🏽

**Review by gomesalexandre — COMMENTED** [review]
Tested with a monkey patch to allow for sub-minimum swap, does what it says on the label, although we will need some changes in web: Currently, the confirm spinner spins until the swap is complete, but the flow for other swaps e.g ZRX is to have the arrow icon change to a spinner while the trade isn't complete:

image

This logic happens here:

https://github.com/shapeshift/web/blob/536c9d00494edd2d81985a0c8be708d692007291/src/components/Trade/TradeConfirm/TradeConfirm.tsx#L154

File: packages/swapper/src/swappers/thorchain/types.ts (hunk 1 above)
Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/types.ts:36]: it really doesnt matter

Comment (0xdef1cafe) [reply packages/swapper/src/swappers/thorchain/types.ts:36]: `[action]` is an array of length exactly 1

Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/types.ts:36]: derp

Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/types.ts]: fixed

File: packages/swapper/src/swappers/thorchain/types.ts [hunk 2]
```diff
@@ -21,6 +21,21 @@ export type PoolResponse = {
   volume24h: string
 }
 
+type MidgardActionOut = {
+  txID: string
+}
+type MidgardAction = {
+  date: string
+  height: string
+  out: [MidgardActionOut]
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/types.ts:30]: is this guaranteed to be an array of exactly length 0?

Comment (0xdef1cafe) [reply packages/swapper/src/swappers/thorchain/types.ts:30]: this is not an opinion or stylistic - `[Type]` is not the same as `Type[]`

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

File: packages/swapper/src/swappers/thorchain/types.ts (hunk 2 above)
Comment (elmutt) [reply packages/swapper/src/swappers/thorchain/types.ts:30]: ah yes good point. long day whoops

Comment (gomesalexandre) [reply packages/swapper/src/swappers/thorchain/types.ts:30]: `MidgardActionsResponse['actions']` has been updated but this is still there. Again, do we want this to be a tuple?

**Review by gomesalexandre — COMMENTED** [review]
Question

**Review by gomesalexandre — APPROVED** [review]
LGTM

Swapped ATOM into LTC and got the spinner while the swap was completing on Thor / green check after it completed

image

**Pull request merged**